Add Code Quailty chapter (#116)
infinum/frontend-handbook: Code Quality/Reviewing a pull request.md

> Optimism is an occupational hazard of programming; feedback is the treatment. - Kent Beck

This guide should help you review code better. A good code review will provide a lot of useful things, but some of them are:

-   a sense of security (that the feature works well),
-   a learning opportunity for both parties,
-   architectural check,
-   code quality check.

## Language of the code review

-   be concise
-   do not ask, advise
-   do not assume the level of knowledge

> -   be open for discussion, but do not bike ßshed

Reviewer: Scharfes S is probably a typo

Author: Fixed. But it was not!

-   recommend fixes instead of pointing out errors

### Good examples:

> this won't work because the function you're calling as `async`.

> this file is palaced in the incorrect folder, it should go to `components/shared`

> I'm not sure this does what you want it to do. Looking at the service as a whole you should probably allow the end user to construct this out of available functions.

### Bad examples:

> can you please update this function documentation as well?

> this is not according to our guidelines

> naming is not correct here

## When do I review pull requests?

Reviewing a pull request could (and should) take time. This is not to be rushed since you vouch for the quality of the code that's on review.

> If you review code for either multiple projects or you work in a larger team you might want to think about assigning a certain fixed part of the day to do them (e.g. I review pull requests in the morning). Everything that comes during the day is to be reviewed first thing tomorrow. This should not include urgent things that block hot fixes, etc.

Reviewer 2: Also worth mentioning that it should be taken into account during estimates?

Author: HM - Good point!

Code review is an integral part of the development cycle and therefore should be time tracked on the appropriate task/service on the project.

## How do I review a pull request?

Start by reading the feature that's delivered. Look at the design files, get to know the business logic behind it, understand the requirements.

Next up is scanning the files for common errors or typos, etc. - things that stand out right away. Comment them and if they are blocking enough return the pull request to the developer.

You should have the project you're reviewing checked out locally and you should run **both** the project and the tests on **every** pull request. Your review is not complete unless you click thru the feature and verify at least the happy path.

Take a look at the code coverage and look for missing things there, are all branched tested or even developed?

You, too, are taking part in the chain of quality assurance here.

## Common things to look out for

> -   the code should have been linted and prettified before the code review

Reviewer 2: This statement is correct, however, should this be the focus of a PR review? Since we're already having tools for formatting, this seems like a distraction to the reviewer.

Author: I agree with you 100%. This should probably go to the creating a PR document instead. Intention with this was "do not review if that's not true".

I'll reword and place this in the creating document as well.

-   the code should be tested using unit tests

> -   the code should not lower the code coverage on the project

Reviewer 2: Worth investigating tools that would verify this? I've seen bots and GH actions for this purpose.

Reviewer: If using karma, you can set up coverage gates so that the build for the branch will fail if the coverage drops below some threshold and this will block merging.

I am not sure if there is a way to check that the coverage dropped compared to target branch, but not low enough to trigger the gate (e.g. master has 82% coverage, gate is 80% and PR drops it to 81%). Gate should also be increased from time to time

Reviewer 2: I know about this, jest also has it. The question is if we want to have a fixed gate or an increasing gate

This guide should help you review code better. A good code review will provide a lot of useful things, but some of them are:

-   a sense of security (that the feature works well),
-   a learning opportunity for both parties,
-   architectural check,
-   code quality check.

## Language of the code review

-   Be concise
-   Do not ask, advise
-   Do not assume the level of knowledge
-   Be open for discussion, but do not bike shed
-   Recommend fixes instead of pointing out errors

### Good examples:

> this won't work because the function you're calling is `async`.

> this file is placed in the incorrect folder, it should go to `components/shared`

> I'm not sure this does what you want it to do. Looking at the service as a whole you should probably allow the end user to construct this out of exported functions.

### Bad examples:

> can you please update this function documentation as well?

> this is not according to our guidelines

> naming is not correct here

> pls fix.

## When do I review pull requests?

Reviewing a pull request could (and should) take time. This is not to be rushed since you vouch for the quality of the code that's on review.

> If you review code for either multiple projects or you work in a larger team, you might want to think about assigning a certain fixed part of the day to do them (e.g. I review pull requests in the morning). Everything that comes during the day is to be reviewed first thing tomorrow. This should not include urgent things that block hot fixes, etc. Communicate this time to your project manager so it can be correctly reflected in your sprint availability.

Reviewer 3: This part is weird. You are talking about assigning a certain part of the day for reviewing pull requests and then in the second part you state that this should be done "first thing tomorrow".

I would go in a different direction. When and how often pull requests should be reviewed is up to you and the project team but as a rule of thumb - a pull request shouldn't be unchecked for more than 24 hours. That means that you should review pull requests at least once a day.
Urgent fixes should, of course, be checked as soon as possible.

Author: Yeah I agree - this was the idea; and the suggestion was to have as a fixed portion of the day (no matter if it's morning or evening or whatever).

I'll reword this a bit.

Author: I cleared it up a bit. Care to give it another read? 758a44f

Reviewer 3: Great, I think you cleared everything up now 👍

> Optimism is an occupational hazard of programming; feedback is the treatment. - Kent Beck

This guide should help you review code better. A good code review will provide a lot of useful things, but some of them are:

-   a sense of security (that the feature works well),
-   a learning opportunity for both parties,
-   architectural check,
-   code quality check.

## Language of the code review

-   Be concise
-   Do not ask, advise
-   Do not assume the level of knowledge

> -   Be open for discussion, but do not bike shed

Reviewer 4: ```suggestion
-   Be open for discussion, but do not bikeshed
```
https://en.wiktionary.org/wiki/bikeshedding

-   Recommend fixes instead of pointing out errors

### Good examples:

> this won't work because the function you're calling is `async`.

> > this file is placed in the incorrect folder, it should go to `components/shared`

Reviewer 5: for this example I think we need an explanation why would it be better to move that file, like 
```
> this file should go to `components/shared` because this is atom component and can be reused on another page
```

Or something like that. If we give an explanation it might open a discussion about it.

Author: Amazing thanks

> > this file should go to `components/shared` because this is atom component and can be reused on another page

Reviewer 4: ```suggestion
> this file should go to `components/shared` because this is an atom component and can be reused on another page
```

> pls fix.

## When do I review pull requests?

Reviewing a pull request could (and should) take time. This is not to be rushed since you vouch for the quality of the code that's on review.

Pull requests should not stay unattended for more than 24 hours. This means you should pick them up and start the review in that period. When you do them however is entirely up to you and your team.

This can be specifically hard to do if you review multiple projects or work in a larger team. If this is the case you might want to think about assigning a certain part of the day to do them.

Be sure to communicate this to your project manager so it gets reflected in your sprint availability.

Code review is an integral part of the development cycle and therefore should be time tracked on the appropriate task/service on the project.

## How do I review a pull request?

Start by reading the feature that's delivered. Look at the design files, get to know the business logic behind it, understand the requirements.

Next up is scanning the files for common errors or typos, etc. - things that stand out right away. Comment them and if they are blocking enough return the pull request to the developer.

> You should have the project you're reviewing checked out locally and you should run **both** the project and the tests on **every** pull request. Your review is not complete unless you click thru the feature and verify at least the happy path.

Reviewer 4: ```suggestion
You should have the project you're reviewing checked out locally and you should run **both** the project and the tests on **every** pull request. Your review is not complete unless you click through the feature and verify at least the happy path.
```
![image](https://user-images.githubusercontent.com/373850/118441748-e88b2380-b6e9-11eb-940c-ba52ddf6f302.png)

> Take a look at the code coverage and look for missing things there, are all branched tested or even developed?

Reviewer 4: ```suggestion
Take a look at the code coverage and look for missing things there, are all branches tested or even developed?
```

You, too, are taking part in the chain of quality assurance here.

## Common things to look out for

-   If the code is not linted or reviewed return the pull request to the developer
-   The code should be tested using unit tests
-   The code should not lower the code coverage on the project
-   The code should follow the design specification
-   Tricky parts of the code that are hard to understand should have either inline documentation or a link to external documentation

> Some of this checks can be automated on pull requests (e.g. code coverage). Talk to your resident devops to help you set this up.

Reviewer 4: ```suggestion
Some of these checks can be automated on pull requests (e.g. code coverage). Talk to your resident devops to help you set this up.
```

## Responsibility

> Every developer is responsible for their code. But - you as a reviewer share the responsibility as well when you sign off a feature and allow merge of it.

Reviewer 4: ```suggestion
Every developer is responsible for their code. But - you as a reviewer share the responsibility as well when you sign off a feature and allow it to merge.
```
